src/encoded/static/components/viz/Workflow/parsing-functions.js

```diff
@@ -4,34 +4,165 @@ import _ from 'underscore';
+ * @property {Node} [outputOf]                      For IO nodes only: Another node reference from which this node is output of, if an output IO node.
+ * @property {Object} [argNamesOnSteps]             For IO nodes only: Mapping of IO name by step name, as IO name varies by step it is described by.
+ * @property {string} [format]                      For IO nodes only: The 'argument_type' from CWL, e.g. 'Workflow Parameter', 'Workflow Output File', etc.
+ */
+
+
+ /**
+ * Type definition for an Edge between nodes.
+ * 
+ * @typedef {Object} Edge
+ * @property {Node} source - Node at which the edge starts.
+ * @property {Node} target - Node at which the edge ends.
+ */
+
+
+
+
+
+/**
+ * Converts a list of steps (each with inputs & outputs) into a list of nodes and list of edges between those nodes to feed directly into the
+ * Workflow graph component.
+ * 
+ * @param {Step[]} analysis_steps                       List of steps from the back-end to generate nodes & edges from.
+ * @param {string} [parsingMethod='output']             Deprecated.
+ * @returns {{ 'nodes' : Node[], 'edges' : Edge[] }}    Container object for the two lists.
+ */
 export function parseAnalysisSteps(analysis_steps, parsingMethod = 'output'){
 
-    /**** Outputs ****/
+    /*************
+     ** Outputs **
+     *************/
 
     var nodes = [];
     var edges = [];
 
-    /* Temp Vars */
-    var ioIdsUsed = {  }; // Keep track of IO arg node ids used, via keys; prevent duplicates by incrementing int val.
+    /*************
+     * Temp Vars *
+     *************/
 
-    /**** Functions ****/
+    var ioIdsUsed = {  };       // Keep track of IO arg node ids used, via keys; prevent duplicates by incrementing int val.
+    var processedSteps = {};    // Keep track of steps already processed & added to graph. We start drawing first step, and its inputs/outputs, then recursively draw steps that its outputs go to. At the end, we restart from step that has not yet been encountered/processed, if any.
 
-    function stepNodeID   (step) { return step.uuid || (step.meta && typeof step.meta === 'object' && step.meta['@id']) || step['@id'] || step.link_id || step.name; }
-    function stepNodeName (step) { return step.display_title || step.title || step.name || step['@id']; }
-    function inputNodeID  (stepInput, readOnly = true) {
-        return preventDuplicateNodeID(
-            stepInput.id /*|| (Array.isArray(stepInput.source) && stepInput.source.length > 0 && stepInput.source[0].name)*/ || stepInput.name,
-            readOnly
-        );
-    }
-    function outputNodeID (stepOutput, readOnly = true){
+    /***************
+     ** Functions **
+     ***************/
+
+     /**
+      * @param {StepIOArgument} stepIOArg - Input or output argument of a step.
+      * @param {boolean} [readOnly=true] - @see preventDuplicateNodeID()
+      * @returns {string} Unique ID for node.
+      */
+    function ioNodeID(stepIOArg, readOnly = true){
```

Comment:
Step (& step nodes) must be unique in their 'name' property (aka steps.name, stepNode.name).
IO arguments *do not* need to be unique on their name _except within a step_ (CWL struct anyway I think) and will auto-generate an additional ID, because there might be multiple steps with same argument name, in addition to multiple files for one argument. Autogenerated ID is just `steps['inputs'||'outputs'].name[ + '~' +
 <duplicate_increment>]`, it might be worthwhile to change this to `steps.name + steps['inputs'||'outputs'].name[ + '~' +
 <duplicate_increment>]`. Doesn't affect graphing, just for semantics/data-completeness/correctness.